scripts/parse.py

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
+
+import argparse
+import json
+
+from jinja2 import Environment
+from jinja2.loaders import FileSystemLoader
+
+
+def main() -> int:
+    parser = argparse.ArgumentParser(
+        prog="AMS swagger parser",
+        description="Extracts AMS configuration data",
+    )
+    parser.add_argument(
+        "swagger_path",
+        help="Path to the swagger json file (Should conform to swagger spec 2.0)",
+    )
+    parser.add_argument(
+        "--output",
+        dest="output_file",
+        help="Destination of the rendered configuration file",
+        default="ams-configuration.md",
+    )
+    args = parser.parse_args()
+    with open(args.swagger_path, mode="r") as f:
+        swagger = json.load(f)
+
+    configs = _parse_config_schema(swagger)
+    nodes = _parse_node_schema(swagger)
+    env = Environment(loader=FileSystemLoader("."))
+    templ = env.get_template("template.md.j2")
+    text = templ.render(configs=configs, nodes=nodes)
+    with open(args.output_file, mode="w+") as op:
+        op.write(text)
+
+
+def _parse_config_schema(swagger) -> dict:
+    ams_config_path = "/1.0/config"
+    config_props = swagger["paths"][ams_config_path]["get"]["responses"]["200"][
+        "schema"
+    ]["properties"]["metadata"]["properties"]["config"]["properties"]
+    for prop in config_props.values():
+        if "description" in prop:
+            prop["description"] = prop["description"].replace("\n", " ")
+    return config_props
+
+
+def _parse_node_schema(swagger) -> dict:
+    node_props = swagger["definitions"]["NodePatch"]["properties"]
+    for base, prop in dict(node_props).items():
+        key = ""
+        if prop["type"] == "object":
+            key = "additionalProperties"
+        if prop["type"] == "array":
+            key = "items"
+        if key:
+            if schema_name := prop[key].get("$ref"):
+                name = schema_name.split("/")[-1]
+                if schema := swagger["definitions"][name]:
+                    for subprop_name, subprop_attrs in schema["properties"].items():
+                        if subprop_name == "id":
+                            continue
+                        new_base = base
+                        if base == "gpus":
+                            new_base = f"{base}.<id>"
+                        node_props[f"{new_base}.{subprop_name}"] = subprop_attrs
+            node_props.pop(base)
+        else:
+            val = node_props.pop(base)
+            node_props[base.replace("_", "-")] = val
+        if "description" in prop:
+            prop["description"] = prop["description"].replace("\n", " ")
+        return node_props
```

Comment:
hmm, is this right?
To me the indentation is wrong, which means we only handle the first prop, then returns the node_props?

Reply:
oh yes good catch thanks. I had refactore the code a bit at the end.